app/services/file-embeddings.ts

```
@@ -0,0 +1,384 @@
+              
+              // Generate content hash
+              const contentHash = createHash('sha256')
+                .update(content)
+                .digest('hex');
+              
+              // Check if we already have an embedding for this exact content
+              const { data: existing } = await supabase
+                .from('file_embeddings')
+                .select('id')
+                .eq('repository_id', dbRepo.id)
+                .eq('file_path', filePath)
+                .eq('content_hash', contentHash)
+                .single();
+              
+              if (existing) {
+                console.log(`Skipping ${filePath} - embedding already exists`);
+                return;
+              }
+              
+              // Prepare content for embedding (truncate if too long)
+              const embeddingContent = prepareContentForEmbedding(filePath, content);
+              
+              // Generate embedding
+              const embedding = await generateEmbedding(embeddingContent);
+              
+              embeddings.push({
+                repository_id: dbRepo.id,
+                file_path: filePath,
+                embedding,
+                content_hash: contentHash,
+                last_indexed_at: new Date().toISOString(),
+              });
+            }
+          } catch (error) {
+            console.error(`Error processing file ${filePath}:`, error);
+          }
+        })
+      );
+      
+      // Add delay to avoid rate limiting
+      if (i + 10 < filePaths.length) {
+        await new Promise(resolve => setTimeout(resolve, 1000));
+      }
+    }
+    
+    // Insert embeddings in batches
+    for (let i = 0; i < embeddings.length; i += 50) {
+      const batch = embeddings.slice(i, i + 50);
+      
+      const { error } = await supabase
+        .from('file_embeddings')
+        .upsert(batch, {
+          onConflict: 'repository_id,file_path',
+        });
+      
+      if (error) {
```

Do you want to continue processing file embeddings if one fails? If so, consider adding retry logic. Or do you want to fail fast if one fails?